### PR TITLE
#158563613 Config options for attribute

### PR DESCRIPTION
**Pull request by missating**
**What does this PR do?**
It shows the configuration options for an additional attribute field when a user clicks on or around the field

**Description of Task to be completed?**

    - click on the attribute field to toggle the show attribute config
    - modify the create category page to have the additional attribute field
    - write the test for both the additional attribute and create the category page
    - add icons for show attribute config

**How should this be manually tested?**

 - `yarn start:dev` to open the web browser on `port 3000` 
-  On the `settings page` when creating `a new category` click on any of the additional attribute field for the configuration optios to be visible

**Any background context you want to provide?**

N/A

**What are the relevant pivotal tracker stories?**

[#158563613](https://www.pivotaltracker.com/story/show/158563613)

**Screenshots (if appropriate)**

Additional attribute field without the config options
screen shot 2018-07-05 at 7 07 16 pm

Additional attribute field with the config options
screen shot 2018-07-05 at 7 07 33 pm

**Questions:**

N/A

### CODE REVIEW

File: src/pages/CreateCategoryPage/index.tsx [hunk 1]
```diff
@@ -5,92 +5,127 @@ import * as React from 'react';
...
+                <img className="form-fields__form-row__draggable-image" src="/images/reorder.svg" />
+                <InputBox
+                  placeholder="Asset tag"
+                  error="This field is required."
+                  // tslint:disable-next-line no-empty
+                  onChange={() => { }}
+                />
+                <SelectBox
+                  classes="create-category-select-width"
+                  options={[{ id: '1', name: 'Textarea' }]}
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: Lines 72, 84 and 96 are repeated. Asides from the repetition, I can envision a scenario where one might need to add more options to the array, say Drop-Down, or Radio Button. Having to make the changes in multiple places isn't scalable. Abstracting this into a separate `fixtures.ts` file and importing it in this file would go a long way in making this code easily maintainable.

File: src/pages/CreateCategoryPage/index.tsx [hunk 2]
```diff
@@ -5,92 +5,127 @@ import * as React from 'react';
...
+    });
+  }
+
+  render() {
+    return (
+      <React.Fragment>
+        <div className="settings-page__heading">
+          <h1 className="settings-page__heading__text">Add a category</h1>
+          <Button
+            alignImageRight={false} name="Preview form" classes="regular--light" icon="/images/previewFormIcon.svg" />
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: I think this would be easier to read if all the attributes are tabbed on a separate line, much like `alignImageRight` is.

File: src/components/AdditionalAttribute/AdditionalAttribute.scss [hunk 3]
```diff
@@ -0,0 +1,30 @@
+@import '../../assets/scss/variables';
+@import 'sass-rem';
+
+.attribute-config-icons {
+  position: relative;
+  left: 534px;
+  display: flex;
+  background-color: map-get($grey, 1200);
+  width: 120px;
```
Comment (Billmike) [inline src/components/AdditionalAttribute/AdditionalAttribute.scss]: No px. Use rem, and change all instances where you used px.

File: src/components/AdditionalAttribute/index.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    {
+      props.showAttributeConfig &&
+      <div className="attribute-config-icons" onClick={props.onClick}>
+        <img src="/images/delete-icon.svg" />
+        <img src="/images/required-icon.svg" />
+        <img src="/images/unique-icon.svg" />
+        <img src="/images/ellipsis-icon.svg" />
+      </div>
+    }
+    <div className={props.showAttributeConfig
```
Comment (Billmike) [inline src/components/AdditionalAttribute/index.tsx]: So with the way this is setup, clicks within this component would always toggle the visibility of the configuration options. I don't think that's the intended functionality. Also, it makes the component impossible to re-use. Could the toggle logic be implemented a little differently? Perhaps the options should only be visible when a valid input type has been selected?

**Comment by andela-kakpobome** [comment]
Don't forget to update the snapshots

File: src/components/Attribute/index.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+          <img className="form-fields__form-row__draggable-image" src="/images/reorder.svg" />
+          <InputBox
+            placeholder="Warranty"
+            // tslint:disable-next-line no-empty
+            onChange={() => { }}
+          />
+          <SelectBox
+            classes="create-category-select-width"
+            name="InputType"
+            value={this.state.value}
```
Comment (Billmike) [inline src/components/Attribute/index.tsx]: While this works for just one input field, it introduces a potential bug when there are multiple input fields. The `onChange()` event-listener would be attached to every instance of the `SelectBox` component, and changing the value of one `Select-Box` would change the value of every `Select-Box` mounted. The reason for this is because they all get their value from the same state tree, so passing the value to each instance of the `SelectBox` component means the onChange event-handler on one `SelectBox` would set the value for all instance of the `SelectBox` to the same thing.

File: src/components/Attribute/index.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    if (classNames.findIndex(name => name === event.target.className.trim()) >= 0) {
+      return;
+    }
+    this.setState({ showAttributeConfig: !this.state.showAttributeConfig });
+  }
+
+  render() {
+    return (
+      <React.Fragment>
+        <div onClick={this.toggleShowAttributeConfig} className={this.state.showAttributeConfig
```
Comment (Billmike) [inline src/components/Attribute/index.tsx]: This works remarkably well. Just one issue though: Clicking the icons should not toggle the `showAttributeConfig` state. That's because future features depend on those icons being visible when clicked on.

File: src/components/SelectBox/interfaces.d.ts [hunk 7]
```diff
@@ -1,6 +1,14 @@
 export interface SelectBoxProps {
   value?: any;
   options: [
+    {
```
Comment (Billmike) [inline src/components/SelectBox/interfaces.d.ts]: Is there any reason why this is defined thrice?

File: src/pages/CreateCategoryPage/index.tsx [hunk 8]
```diff
@@ -5,92 +5,111 @@ import * as React from 'react';
...
+// fixtures
+import { inputTypes } from './fixtures';
 
-        <p className="category-name__text">Category Name</p>
+/**
+ * This method returns the input component with the placeholder and error passed as argument
+ * 
+ * @returns {component}
+ */
+function renderInputBox(placeholder, error) {
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: Use arrow function/closure over the actual `function` keyword

File: src/pages/CreateCategoryPage/index.tsx [hunk 9]
```diff
@@ -5,92 +5,111 @@ import * as React from 'react';
...
+/**
+ * This method returns the input component with the placeholder and error passed as argument
+ * 
+ * @returns {component}
+ */
+const renderInputBox = (placeholder, error) => {
+  return (
+    <InputBox
+      placeholder={placeholder}
+      error={error || ''}
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: Rather than make use of the `||` operator, you could make use of ES6 default parameters to make your functions easily readable. So, instead of a conditional here, you could re-write your function like so:
```javascript
const renderInputBox = (placeholder, error = '') => {
  return (<InputBoxJSX
                  placeholder={placeholder}
                  error={error}
                  onChange={() => { }}
               />);
}
```
This way, you don't have to pass in a value for error is you don't need to.

File: src/pages/CreateCategoryPage/index.tsx [hunk 10]
```diff
@@ -5,92 +5,111 @@ import * as React from 'react';
...
-            <img className="form-fields__form-row__draggable-image" src="/images/reorder.svg" />
-            <InputBox
-              placeholder="Asset tag"
-              // tslint:disable-next-line no-empty
-              onChange={() => {}}
-            />
-            <SelectBox
+          <p className="category-name__text">Category Name</p>
+
+          {renderInputBox('Apple Tv', '')}
```
Comment (Billmike) [inline src/pages/CreateCategoryPage/index.tsx]: Once your `renderInputBox` function is updated, you can take out the `''` here.

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 11]
```diff
@@ -49,13 +49,24 @@ describe('AddAssetPane component', () => {
...
   });
+
+  it('should show the addAssetForm when the contiue button is clicked', () => {
+    const wrapper = shallow(<AddAssetPane {...props} />);
+
+    wrapper.find('#continue-asset-button').simulate('click');
+
+    expect(props.onClick).toHaveBeenCalled();
+    expect(wrapper.instance().state.showAssetForm).toBeTruthy();
+    expect(wrapper.find('p').length).toBe(1);
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: This assertion is too generic, either remove it or make it more specific.

File: src/components/AddAssetPane/AddAssetPane.test.tsx [hunk 12]
```diff
@@ -49,13 +49,24 @@ describe('AddAssetPane component', () => {
...
       target: {
         value: 'Chairs',
       },
     });
+
     expect(wrapper.state().value).toEqual('Chairs');
   });
+
+  it('should show the addAssetForm when the contiue button is clicked', () => {
+    const wrapper = shallow(<AddAssetPane {...props} />);
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/AddAssetPane.test.tsx]: Space around the `...props`

File: src/components/AddAssetPane/index.tsx [hunk 13]
```diff
@@ -92,12 +92,17 @@ class AddAssetPane extends React.Component<AddAssetPaneProps> {
...
           name="CONTINUE"
           disabled={this.state.value === 'Asset Category' ? true : false}
           size={`${this.changeButtonClass(this.state.value)} blue`}
           onClick={this.toggleAssetFormVisibility}
         />
-        <Button id="cancel-asset-button" name="CANCEL" size="regular white" onClick={this.props.onClick} />
+        <Button 
+          id="cancel-asset-button" 
+          name="CANCEL" 
+          size="regular white"
```
Comment (andela-kakpobome) [inline src/components/AddAssetPane/index.tsx]: For semantic reasons, it'll be better to pass `white` as a `type` prop, not size. We shouldn't even need to have a `white` style since we already have hollow, the background would be transparent, hence the white of the form would be the color that button would assume.

File: src/components/Attribute/Attribute.test.tsx [hunk 14]
```diff
@@ -0,0 +1,42 @@
+// react libraries
+import * as React from 'react';
+
+// third party libraries
+import { shallow } from 'enzyme';
+
+// components
+import AdditionalAttribute from './';
```
Comment (andela-kakpobome) [inline src/components/Attribute/Attribute.test.tsx]: Shouldn't this be `Attribute` now?

File: src/components/Attribute/Attribute.test.tsx [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+    });
+
+    expect(wrapper.state().value).toEqual('Datepicker');
+  });
+
+  it('should show the config options for attribute field when showAttributeConfig is true', () => {
+    const wrapper = shallow(<AdditionalAttribute />);
+
+    wrapper.find('div.form-fields__form-row').simulate('click', { 
+      target: {
```
Comment (andela-kakpobome) [inline src/components/Attribute/Attribute.test.tsx]: What's the value of passing a target here?

File: src/components/Attribute/index.tsx [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    this.setState({ showAttributeConfig: !this.state.showAttributeConfig });
+  }
+
+  render() {
+    const attributeFieldArray = [
+      { id: '2', name: 'Datepicker' },
+      { id: '3', name: 'Checkbox' },
+      { id: '4', name: 'Radio Button' },
+    ];
+    const attributeFieldOptions = attributeFieldArray.map((field) => {
```
Comment (andela-kakpobome) [inline src/components/Attribute/index.tsx]: There's no difference between this and `const attributeFieldOptions = [...attributeFieldArray]`, what's the point of it?

File: src/components/SelectBox/interfaces.d.ts [hunk 17]
```diff
@@ -4,8 +4,7 @@ export interface SelectBoxProps {
     {
       id: string,
       name: string
-    }
-  ];
+    }];
```
Comment (andela-kakpobome) [inline src/components/SelectBox/interfaces.d.ts]: Why this change? The former is preferred

File: src/pages/CreateCategoryPage/index.tsx [hunk 18]
```diff
@@ -5,92 +5,108 @@ import * as React from 'react';
...
-              onChange={() => {}}
-            />
-            <SelectBox
-              classes="create-category-select-width"
-              options={[{ id: '1', name: 'Textarea' }]}
-            />
+
+          <h2 className="additional-fields__section-title">
+            Additional Fields
+            </h2>
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: Improper indentation

File: src/pages/CreateCategoryPage/index.tsx [hunk 19]
```diff
@@ -5,92 +5,108 @@ import * as React from 'react';
...
+          <Attribute />
 
-        <a className="additional-fields__new-attribute" href="">
-          <img className="additional-fields__new-attribute__icon" src="/images/addIcon.svg" alt="" />
-          Add New Attribute
-        </a>
+          <a className="additional-fields__new-attribute" href="">
+            <img className="additional-fields__new-attribute__icon" src="/images/addIcon.svg" alt="" />
+            Add New Attribute
+            </a>
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: Improper indentation

File: src/pages/CreateCategoryPage/index.tsx [hunk 20]
```diff
@@ -5,92 +5,108 @@ import * as React from 'react';
...
+              <h4 className="form-fields__heading__attribute-name">Attribute name</h4>
+              <h4 className="form-fields__heading__input-type">Input type</h4>
+            </div>
+            <div className="form-fields__form-row form-fields__form-row--error">
+              <img className="form-fields__form-row__draggable-image" src="/images/reorder.svg" />
+              {renderInputBox('Asset tag', 'This field is required.')}
+              <SelectBox
+                classes="create-category-select-width"
+                options={showFieldType as any}
+              />
```
Comment (andela-kakpobome) [inline src/pages/CreateCategoryPage/index.tsx]: The whole point of creating this function was to ensure that the codebase is DRY, but your implementation still allows duplication(see 63, 71, 79), hence is cosmetic at best. 

That being said, why don't you just use the attribute component here instead of creating separate input and select boxes

File: src/components/AddAssetPane/index.tsx [hunk 21]
```diff
@@ -92,12 +92,17 @@ class AddAssetPane extends React.Component<AddAssetPaneProps> {
 
       <div className="asset-pane__buttons">
         <Button
+          id="continue-asset-button"
           name="CONTINUE"
           disabled={this.state.value === 'Asset Category' ? true : false}
```
Comment (Dammyy) [inline src/components/AddAssetPane/index.tsx]: can we have this instead? `disabled={this.state.value === 'Asset Category'}` by default this returns a boolean, no need for the extra condition `? true : false` .

File: src/components/Attribute/index.tsx [hunk 22]
```diff
@@ -0,0 +1,82 @@
...
+      return field;
+    });
+
+    return (
+      <React.Fragment>
```
Comment (kevinejiro) [inline src/components/Attribute/index.tsx]: There isn't any need to wrap this in `<React.Fragment>`

File: src/components/Attribute/index.tsx [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+          <div className="attribute-config-icons">
+            <img src="/images/delete-icon.svg" />
+            <img src="/images/required-icon.svg" />
+            <img src="/images/unique-icon.svg" />
+            <img src="/images/ellipsis-icon.svg" />
+          </div>
+
+          <img className="form-fields__form-row__draggable-image" src="/images/reorder.svg" />
+          <InputBox
+            placeholder="Warranty"
```
Comment (kevinejiro) [inline src/components/Attribute/index.tsx]: Shouldn't the placeholder be something more generic like `Attribute Name`

File: src/components/Attribute/Attribute.test.tsx [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+    const attributeContainer = document.createElement('div');
+    document.body.appendChild(attributeContainer);
+
+    const wrapper = mount(<Attribute { ...props } />, { attachTo: attributeContainer });
+
+    wrapper.find('.form-fields__form-row').simulate('click');
+  
+    simulant.fire(document.body, 'click');
+
+    // TODO: find out how to write assertion.
```
Comment (kevinejiro) [inline src/components/Attribute/Attribute.test.tsx]: Is this a personal comment ? If so why is this in the general codebase.

File: src/components/Attribute/index.tsx [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+   * @returns {void}
+   */
+  hideAttributeConfig = () => {
+    this.setState({
+      showAttributeConfig: false,
+    });
+  }
+
+  render() {
+    const showFieldType = inputTypes.map((field) => {
```
Comment (Billmike) [inline src/components/Attribute/index.tsx]: We already established that mapping here is irrelevant. You can just pass the inputTypes directly to the options props of the SelectBox.

File: src/components/Attribute/Attribute.test.tsx (hunk 24 above)
Comment (missating) [reply src/components/Attribute/Attribute.test.tsx]: No, It's not. This PR is a WIP because I'm still looking for ways to fix the assertion for that test.

File: src/components/Attribute/Attribute.test.tsx [hunk 26]
```diff
@@ -0,0 +1,79 @@
...
+    });
+
+    document.getElementById = (id) => {
+      return {
+        contains: arg => false,
+      } as any;
+    };
+
+    const wrapper = mount(<Attribute  { ...props } />);
+    expect(wrapper.find('.form-fields__form-row--active').length).toBe(0);
```
Comment (sccofield) [inline src/components/Attribute/Attribute.test.tsx]: Add a line before this.

File: src/components/Attribute/index.tsx [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+   */
+  hideAttributeConfig = () => {
+    this.setState({
+      showAttributeConfig: false,
+    });
+  }
+
+  render() {
+    return (
+        <div id={this.props.id}
```
Comment (sccofield) [inline src/components/Attribute/index.tsx]: Check your indentation here.

**Comment by andela-kakpobome** [comment]
@missating The the snapshots were not updated here
